pigosat.go

```diff
@@ -233,6 +234,172 @@ func (p *Pigosat) AddClauses(clauses Formula) {
+//   p.Assume(-2)               // additionally assume unit clause Clause{-2, 0}
+//   res, solution := p.Solve() // assumes 1 and -2 to hold
+//
+//   if res == Unsatisfiable {
+//       if p.FailedAssumption(1) {
+//           // unit clause Clause{1, 0} was necessary to derive Unsatisfiable
+//       }
+//       if p.FailedAssumption(-2) {
+//           // unit clause Clause{-2, 0} was necessary to derive Unsatisfiable
+//       }
+//       // at least one but also both could be necessary
+//
+//       p.Assume(17)  // Previous assumptions are removed. Now assume unit
+//                     // clause Clause{17, 0} for the next call to Solve().
+//
+//   } else if res == Satisfiable {
+//       // The assumptions 1 and -2 hold, so the follow assertion succeeds.
+//       if !(solution[1] && !solution[2]) {
+//           panic("assertion failed")
+//       }
+//
+//       // Assumptions valid, as always, until calls to AddClauses(), etc.
+//       // Further, when entering Solve() the solver knows that the previous
+//       // call returned SAT and it can safely reset the previous assumptions.
+//   } else  {
+//       // res == Unknown
+//
+//       // Assumptions valid, but assignment invalid except for top level
+//       // assigned literals which necessarily need to have this value if the
+//       // formula is SAT
+//
+//       // As above the solver knows that the previous call returned Unknown
+//       // and will reset assumptions before doing anything else.
+//   }
+func (p *Pigosat) Assume(lit Literal) {
+	defer p.ready(false)()
+	C.picosat_assume(p.p, C.int(lit))
+}
+
+// Returns non zero if the literal is a failed assumption, which is defined as
+// an assumption used to derive unsatisfiability. This is as accurate as
+// generating core literals, but still of course is an overapproximation of the
+// set of assumptions really necessary. The technique does not need clausal core
+// generation nor tracing to be enabled and thus can be much more effective. The
+// function can only be called as long the current assumptions are valid. See
+// Assume() for more details.
+func (p *Pigosat) FailedAssumption(lit Literal) bool {
+	defer p.ready(true)()
+	// picoast_failed_assumption SIGABRTs if the following conditional is true
+	if p.Res() != Unsatisfiable || lit == 0 {
+		return false
+	}
+	return C.picosat_failed_assumption(p.p, C.int(lit)) != 0
+}
+
+// Returns a list of failed assumption in the last call to Solve(). It only
+// makes sense if the last call to Solve() returned Unsatisfiable.
+func (p *Pigosat) FailedAssumptions() []Literal {
+	defer p.ready(true)()
+	if p.Res() != Unsatisfiable {
+		return nil
+	}
+
+	litPtr := C.picosat_failed_assumptions(p.p)
+	return p.litArrayToSlice(litPtr)
+}
+
+// Converts a 0-terminated array of literal results to a slice.
+// Does not acquire internal locks.
+func (p *Pigosat) litArrayToSlice(litPtr *C.int) []Literal {
+	if litPtr == nil || *litPtr == 0 {
+		return []Literal{}
```

Comment:
From the way I read the documentation for `picosat_failed_assumptions`:
```
/* Returns a zero terminated list of failed assumption in the last call to
 * 'picosat_sat'.  The pointer is valid until the next call to
 * 'picosat_sat' or 'picosat_failed_assumptions'.  It only makes sense if the
 * last call to 'picosat_sat' returned 'PICOSAT_UNSATISFIABLE'.
 */
```
 I think it would take an internal picosat error for `picosat_failed_assumptions` to return a NULL pointer. Perhaps it makes more sense to panic when `litPtr == nil`?